coderd/agentapi/audit.go

```diff
@@ -0,0 +1,117 @@
+package agentapi
+
+import (
+	"context"
+	"encoding/json"
+	"net/http"
+	"strconv"
+	"sync/atomic"
+
+	"github.com/google/uuid"
+	"golang.org/x/xerrors"
+	"google.golang.org/protobuf/types/known/emptypb"
+
+	"cdr.dev/slog"
+
+	agentproto "github.com/coder/coder/v2/agent/proto"
+	"github.com/coder/coder/v2/coderd/audit"
+	"github.com/coder/coder/v2/coderd/database"
+	"github.com/coder/coder/v2/codersdk/agentsdk"
+)
+
+type AuditAPI struct {
+	AgentFn  func(context.Context) (database.WorkspaceAgent, error)
+	Auditor  *atomic.Pointer[audit.Auditor]
+	Database database.Store
+	Log      slog.Logger
+}
+
+func (a *AuditAPI) ReportConnection(ctx context.Context, req *agentproto.ReportConnectionRequest) (*emptypb.Empty, error) {
+	// We will use connection ID as request ID, typically this is the
+	// SSH session ID as reported by the agent.
+	connectionID, err := uuid.FromBytes(req.GetConnection().GetId())
+	if err != nil {
+		return nil, xerrors.Errorf("connection id from bytes: %w", err)
+	}
```

Comment:
What happens to the connection request if we fail to report? Do we fail-open or fail-closed?

Reply:
This UUID must be provided by the agent, if this fails it's a developer error.

An argument allowing the DB fetches for additional context to fail could be made though (we could make it best effort).

Reply:
My question was more related to the general case of what should happen to the connection request if we fail to audit log the request, but I think it may be outside the scope of this PR.

Reply:
Ah, it was decided to allow requests. Best-effort auditing so that we don't delay connection attempts.

Reply:
Got it, thanks!